### PR TITLE
Core role managment

### PR DESCRIPTION
**Pull request by janm-dw**

### CODE REVIEW

File: core/coreobjects/include/coreobjects/property_object.h [hunk 1]
```diff
@@ -329,6 +331,12 @@ DECLARE_OPENDAQ_INTERFACE(IPropertyObject, IBaseObject)
      * holds an event args object that contains a list of properties updated.
      */
     virtual ErrCode INTERFACE_FUNC getOnEndUpdate(IEvent** event) = 0;
+
+    /*!
+     * @brief Gets the permission manager of component.
+     * @param[out] permissionManager The permission manager of component.
```
Comment (JakaMohorkoDS) [inline core/coreobjects/include/coreobjects/property_object.h:337]:
```suggestion
     * @brief Gets the permission manager of property object.
     * @param[out] permissionManager The permission manager of property object.
```

File: core/coreobjects/src/CMakeLists.txt [hunk 2]
```diff
@@ -268,6 +268,7 @@ rtgen_config(${PROJECT_NAME}
 
 target_link_libraries(${LIB_NAME} PUBLIC daq::coretypes
                                          tsl::ordered_map
+                                         daq::security
```
Comment (JakaMohorkoDS) [inline core/coreobjects/src/CMakeLists.txt:271]: Security should then be in coreobjects. At least the permission manager part. Coreobjects should not link any opendaq library.

File: core/opendaq/security/include/opendaq/authentication_provider.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+{
+    /*!
+     * @brief Authenticate user using username and password. If authentication is successful, a User instance is returned.
+     * Otherwise an exception is thrown.
+     * 
+     * @param username The username.
+     * @param password The password in plain text.
+     * @param user[out] And instance of successfully authenticated user. If authentication is not successful, an exception is thrown.
+     */
+    virtual ErrCode INTERFACE_FUNC authenticate(IString* usernanme, IString* password, IUser** userOut) = 0;
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/authentication_provider.h:47]:
```suggestion
    virtual ErrCode INTERFACE_FUNC authenticate(IString* username, IString* password, IUser** userOut) = 0;
```

File: core/opendaq/security/include/opendaq/permission_config.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * @ingroup opendaq_security
+ * @addtogroup opendaq_security_permission_config PermissionConfig
+ * @{
+ */
+
+/*!
+ * @brief A class which describes a permission configuration for openDAQ object.A configuration object
+ * can be constructed using the permission builder class.
+ */
+DECLARE_OPENDAQ_INTERFACE(IPermissionConfig, IBaseObject)
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_config.h:34]: I'd avoid `Config` in interface names, as it clashes with names of other interfaces. I'd just name this `IPermission`, or something like `IPermissionParams`/`IPermissionInfo`.

File: core/opendaq/security/include/opendaq/permission_manager.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * @ingroup opendaq_security
+ * @addtogroup opendaq_security_permission_manager PermissionManager
+ * @{
+ */
+
+
+/*!
+ * @brief Enumeration of available access permissions
+ */
+enum Permission
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_manager.h:35]:
```suggestion
enum class Permission : uint32_t
```

File: core/opendaq/security/include/opendaq/permission_manager_factory.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * @addtogroup opendaq_security_permission_manager Factories
+ * @{
+ */
+
+/*!
+ * @brief Creates a default permission manager object.
+ */
+inline PermissionManagerPtr PermissionManager()
+{
+    PermissionManagerPtr obj(createWithImplementation<IPermissionManager, PermissionManagerImpl>(nullptr));
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_manager_factory.h:33]: We should have factories defined in the interface header for usage outside of the C++ code/in bindings.

File: core/opendaq/security/include/opendaq/permission_manager_private.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#pragma once
+#include <opendaq/permission_manager.h>
+#include <coretypes/dictobject.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+/*!
+ * @brief Provides access to private methods of the permission manager.
+ */
+DECLARE_OPENDAQ_INTERFACE(IPermissionManagerPrivate, IBaseObject)
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_manager_private.h:26]: Is this meant to be accessed my module developer, or is it openDAQ core-only? If the latter is true, we should name this `IPermissionManagerInternal`.

File: core/opendaq/security/include/opendaq/user_private.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+#pragma once
+#include <opendaq/user.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+/*!
+ * @brief Provides access to private methods of the user interface.
+ */
+DECLARE_OPENDAQ_INTERFACE(IUserPrivate, IBaseObject)
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/user_private.h:25]: Likely should be `IUserInternal`

File: core/opendaq/security/src/permission_manager_impl.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    config = builder.build();
+
+    updateChildPermissions();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode INTERFACE_FUNC PermissionManagerImpl::isAuthorized(IUser* user, Permission permission, Bool* authorizedOut)
+{
+    OPENDAQ_PARAM_NOT_NULL(authorizedOut);
+    *authorizedOut = false;
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/src/permission_manager_impl.cpp:40]: What happens if a device vendor does not configure the permission manager? We should at some point probably initialize some component to allow everything. I'd maybe set up the permission configuration of components that have a `nullptr` parent to allow everything.

File: core/opendaq/security/include/opendaq/permission_config_builder.h [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    // [returnSelf]
+    /*!
+     * @brief Configure component to inherit or ignore permissions from parent object.
+     * @param inherit Flag if components should inherit permissions from it's parent object.
+     */
+    virtual ErrCode INTERFACE_FUNC inherit(Bool inherit) = 0;
+
+    // [returnSelf]
+    /*!
+     * @brief Set prmissions for given group.
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_config_builder.h:46]:
```suggestion
     * @brief Set permissions for given group.
```

File: core/opendaq/security/include/opendaq/permission_config_builder.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+     * @param groupId The id of a group to set permissions for.
+     * @param permissionFlags Bit mask of permissions to allow or deny.
```
Comment (JakaMohorkoDS) [inline core/opendaq/security/include/opendaq/permission_config_builder.h:48]: I'm maybe missing a comment that links the `permissions` enum to the bit mask. Basically some guideline on how to construct the bitmask.

**Review by JakaMohorkoDS — COMMENTED** [review]
Aside from some minor stylistic comments that we should likely discuss, the implementation looks clean and good to me!

File: core/coreobjects/include/coreobjects/property_impl.h [hunk 12]
```diff
@@ -202,9 +206,13 @@ class PropertyImpl : public ImplementationOf<IProperty, ISerializable, IProperty
         : PropertyImpl(name, BaseObjectPtr(defaultValue), true)
     {
         this->valueType = ctObject;
+
         if (defaultValue == nullptr)
             this->defaultValue = PropertyObject().detach();
 
+        auto objPermissionManager = this->defaultValue.asPtr<IPropertyObject>().getPermissionManager().asPtr<IPermissionManagerPrivate>();
```
Comment (dejancrn) [inline core/coreobjects/include/coreobjects/property_impl.h:213]: Use borrow casting.

asPtr<Interface>(**true**);

File: core/opendaq/security/include/opendaq/authentication_provider_factory.h [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+{
+    AuthenticationProviderPtr obj(createWithImplementation<IAuthenticationProvider, StaticAuthenticationProviderImpl>(List<IUser>()));
+    return obj;
+}
+
+/*!
+ * @brief Creates an authentication provider out of static list of users.
+ * @param users List of user objects.
+ */
+inline AuthenticationProviderPtr StaticAuthenticationProvider(const ListPtr<IUser>& users)
```
Comment (dejancrn) [inline core/opendaq/security/include/opendaq/authentication_provider_factory.h:42]: Any particular reason why we use inline factories? This means that implementation files (cpp or static libs) have to be used when creating providers, unless the object is header only, which is not.

File: core/opendaq/security/include/opendaq/authentication_provider.h (hunk 3 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/authentication_provider.h]: Fixed

File: core/opendaq/security/include/opendaq/permission_config_builder.h (hunk 10 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/permission_config_builder.h]: Fixed

File: core/opendaq/security/include/opendaq/permission_config_builder.h (hunk 11 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/permission_config_builder.h]: Updated

File: core/opendaq/security/include/opendaq/permission_manager.h (hunk 5 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/permission_manager.h:35]: Ommiting class keyword was intentional because of implicit conversion to Int.
For example, when using permission config builder

```
PermissionConfigBuilder().set("admin", Permission::Read | Permission::Write).set("user", Permission::Read);
```

now becomes 

```
PermissionConfigBuilder().set("admin", (Int) Permission::Read | Permission::Write).set("user", (Int) Permission::Read);
```

Comment (JakaMohorkoDS) [reply core/opendaq/security/include/opendaq/permission_manager.h:35]: I know, it's just that we're using enum classes everywhere in the SDK, and doing explicit conversions. I think it might be an issue if we're passing them to interfaces, but I'm not 100% sure on that. 

@Unsttopabull Do you think it's okay if we use non-class enumerations here?

File: core/opendaq/security/src/permission_manager_impl.cpp (hunk 9 above)
Comment (janm-dw) [reply core/opendaq/security/src/permission_manager_impl.cpp:40]: I added a default group "everyone" and allow all permissions for it on the root property object.

File: core/opendaq/security/include/opendaq/authentication_provider_factory.h (hunk 13 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/authentication_provider_factory.h:41]: No particular reasons. I have moved the factories to cpp files.

File: core/opendaq/security/include/opendaq/permission_manager.h (hunk 5 above)
Comment (janm-dw) [reply core/opendaq/security/include/opendaq/permission_manager.h]: I have refactored a builder to use a list of Permission values instead.

File: core/corecontainers/include/coretypes/listobject_factory.h [hunk 14]
```diff
@@ -33,13 +33,21 @@ using ListPtr = ListObjectPtr<IList, T, S>;
 template <class T, class S = typename InterfaceOrTypeToSmartPtr<T>::SmartPtr, class U>
 static void addToList(ListPtr<T, S>& list, U&& element)
 {
-    list.unsafePushBack(std::forward<U>(element));
+    if constexpr (std::is_enum_v<U>)
```
Comment (janm-dw) [inline core/corecontainers/include/coretypes/listobject_factory.h:36]: @Unsttopabull this is a temporary workaround to support enums in ListObject

Comment (janm-dw) [reply core/corecontainers/include/coretypes/listobject_factory.h]: I have reverted this change, because we now use permission mask builder instead of an list object

**Review by JakaMohorkoDS — APPROVED** [review]
Lgtm, but I did not yet do a really detailed review after the last iteration. We'll likely keep on improving the model it's integrated into server/client modules anyhow.
